app/upgrades/v018/upgrades.go

```diff
@@ -0,0 +1,153 @@
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	govv1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+	paramstypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	ibctransfertypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	exported "github.com/cosmos/ibc-go/v7/modules/core/exported"
+)
+
+func CreateV18UpgradeHandler(mm *module.Manager, configurator module.Configurator, keepers *keepers.AppKeepers) upgradetypes.UpgradeHandler {
+	return func(ctx sdk.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+		logger := ctx.Logger().With("upgrade", UpgradeName)
+
+		ctx.Logger().Info(`
+			; 
+			;;
+			;';. 
+			;  ;;
+			;   ;;
+			;    ;;
+			;    ;;
+			;   ;'
+			;  ' 
+		,;;;,; 
+		;;;;;;
+		;;;;; 
+
+		##     ##      #####          ##    #######  
+		##     ##     ##   ##       ####   ##     ## 
+		##     ##    ##     ##        ##   ##     ##
+		##     ##    ##     ##        ##    ####### 
+	 	 ##   ##     ##     ##        ##   ##     ##
+		  ## ##       ##   ##  ###    ##   ##     ##
+		   ###         #####   ###  ######  ####### 
+	
+			; 
+			;;
+			;';. 
+			;  ;;
+			;   ;;
+			;    ;;
+			;    ;;
+			;   ;'
+			;  ' 
+		,;;;,; 
+		;;;;;;
+		;;;;; 
+				
+		`)
+
+		// https://github.com/cosmos/cosmos-sdk/pull/12363/files
+		// Set param key table for params module migration
+		for _, subspace := range keepers.ParamsKeeper.GetSubspaces() {
+			subspace := subspace
+
+			var keyTable paramstypes.KeyTable
+			switch subspace.Name() {
+			case authtypes.ModuleName:
+				keyTable = authtypes.ParamKeyTable() //nolint:staticcheck
+			case banktypes.ModuleName:
+				keyTable = banktypes.ParamKeyTable() //nolint:staticcheck
+			case stakingtypes.ModuleName:
+				keyTable = stakingtypes.ParamKeyTable() //nolint:staticcheck
+			case distrtypes.ModuleName:
+				keyTable = distrtypes.ParamKeyTable() //nolint:staticcheck
+			case minttypes.ModuleName:
+				keyTable = minttypes.ParamKeyTable()
+			case slashingtypes.ModuleName:
+				keyTable = slashingtypes.ParamKeyTable() //nolint:staticcheck
+			case govtypes.ModuleName:
+				keyTable = govv1.ParamKeyTable() //nolint:staticcheck
+			case crisistypes.ModuleName:
+				keyTable = crisistypes.ParamKeyTable() //nolint:staticcheck
+			// ibc types
+			case ibctransfertypes.ModuleName:
+				keyTable = ibctransfertypes.ParamKeyTable()
+			// bitsong modules
+			case fantokentypes.ModuleName:
+				keyTable = fantokentypes.ParamKeyTable() //nolint:staticcheck
+				if !subspace.HasKeyTable() {
+					subspace.WithKeyTable(keyTable)
+				}
+			// wasmd module
+			case wasmtypes.ModuleName:
+				keyTable = wasmtypes.ParamKeyTable() //nolint:staticcheck
+			default:
+				continue
```

Comment:
do we have other modules to migrate? maybe add `logger.Info()` if we do not migrate something?

Reply:
No. The only custom module bitsong has after this migration is the fantoken. The minimum sdk modules migrated were referenced here: 
- https://github.com/cosmos/cosmos-sdk/blob/0f99829fabe34ae489d71a8524fb4c82632965e3/simapp/upgrades.go 

here are some additional references for modules needed migation:
- osmosis: https://github.com/osmosis-labs/osmosis/blob/main/app/upgrades/v21/upgrades.go#L73
- juno: https://github.com/CosmosContracts/juno/blob/main/app/upgrades/v16/upgrades.go#L55